README.md
# griddle

Griddle _simplifies_ the concept of creating 2D games or UI applications within
a 2D-matrix, or _grid_, which in turn makes it a suitable cross-platform
graphics framework for lower fidelity games or apps.

[![On pub.dev][pub_img]][pub_url]
[![Code coverage][cov_img]][cov_url]
[![Github action status][gha_img]][gha_url]
[![Dartdocs][doc_img]][doc_url]
[![Style guide][sty_img]][sty_url]

[pub_url]: https://pub.dartlang.org/packages/griddle
[pub_img]: https://img.shields.io/pub/v/griddle.svg
[gha_url]: https://github.com/matanlurey/griddle/actions
[gha_img]: https://github.com/matanlurey/griddle/workflows/Dart/badge.svg
[cov_url]: https://codecov.io/gh/matanlurey/griddle
[cov_img]: https://codecov.io/gh/matanlurey/griddle/branch/main/graph/badge.svg
[doc_url]: https://www.dartdocs.org/documentation/griddle/latest
[doc_img]: https://img.shields.io/badge/Documentation-griddle-blue.svg
[sty_url]: https://pub.dev/packages/neodart
[sty_img]: https://img.shields.io/badge/style-neodart-9cf.svg

It is _inspired_ by:

- [`tcell`, a Go package that provides cell-based views for terminals][tcell]
- [`termbox`, a minimalistic API to write text-based UIs][termbox]
- [`termpixels`, the terminal as a character-cell matrix][termpixels]

[tcell]: https://github.com/gdamore/tcell
[termbox]: https://github.com/nsf/termbox-go
[termpixels]: https://github.com/loganzartman/termpixels

## Purpose

Creating simple 2D programs that run inside a terminal (or terminal emulator) is
complicated. The goal of `griddle` is to **abstract a terminal-like screen into
a 2D _grid_ of character cells**.

Like [`termpixels`][termpixels], this project makes the terminal more
accessible and more fun, but in Dart!

To learn more about `griddle`, read our [design philosophy][design.md].

[design.md]: DESIGN.md

## Usage

![Example app running](https://user-images.githubusercontent.com/168174/189504284-4e09879e-75bc-4916-afe0-998f1fa0e5ae.gif)

```dart
import 'dart:io' show stdout;
import 'dart:math' as math;

import 'package:griddle/griddle.dart';

void main() {
  final screen = Screen.display(Display.fromAnsiTerminal(
    stdout,
    width: () => stdout.terminalColumns,
    height: () => stdout.terminalLines,
  ));

  const string = 'Hello World, from Griddle for Dart!';

  Stream<void>.periodic.listen((_) {
    screen.clear();

    for (var i = 0; i < string.length; i++) {
      final t = DateTime.now().millisecondsSinceEpoch / 1000;
      final f = i / string.length;
      final c = Color.fromHSL(f * 300 + t, 1, 0.5);
      final x = screen.width ~/ 2 - string.length ~/ 2;
      final o = math.sin(t * 3 + f * 5) * 2;
      final y = (screen.height / 2 + o).round();

      screen.print(string[i], x + i, y, foreground: c);
    }

    screen.update();
  });
}
```

(For the full example, see [example/example.dart](example/example.dart))

## Contributing

**This package welcomes [new issues][issues] and [pull requests][fork].**

[issues]: https://github.com/matanlurey/griddle/issues/new
[fork]: https://github.com/matanlurey/griddle/fork

Changes or requests that do not match the following criteria will be rejected:

1. Common decency as described by the [Contributor Covenant][code-of-conduct].
2. Making this library brittle.
3. Adding platform-specific functionality.
4. A somewhat arbitrary bar of "complexity", everything should be _easy to use_.

[code-of-conduct]: https://www.contributor-covenant.org/version/1/4/code-of-conduct/


CHANGELOG.md
# CHANGELOG

## 0.5.0

- **BREAKING**: `Display.fromAnsiTerminal` hides the cursor by default.

  In order to show the cursor either:

  - Set `hideCursor: false` in the constructor.
  - Call `<Display>.showCursor()`.

- **BREAKING**: Added `<Display>.close`, which may optionally performn cleanup.

  For `Display.fromAnsiTerminal`, `close` resets the cursor (make it visible).

- Added `<Display>.hideCursor`, `<Display>.showCursor`.

## 0.4.0

- **BREAKING**: `Display.fromAnsiTerminal` reports `height` as `height - 1`.

  This is because attempting to write the entire height of the terminal causes
  scrollback in most terminals, which in practice means the first line of every
  UI is accidentally hidden.

- **BREAKING**: Split `Buffer` into two interfaces: `Buffer` (immutable) and
  `WritableBuffer`.

  - If you were just using `Buffer` indirectly through `Screen`, no API changes.
  - If you were creating a `Buffer` and writing to it, use `WritableBuffer` now:

  ```diff
  - final buffer = Buffer(3, 3);
  + final buffer = WritableBuffer(3, 3);

  buffer.print('Hello', 0, 1);
  ```

- Added `<WritableBuffer>.fillFrom`, which is _spiritually_ a [`BITBLT`][]:

  ```dart
  // Pre-define a sprite-like object in an immutable buffer.
  final other = Buffer.fromMatrix([
    [Cell('┏'), Cell('┓')],
    [Cell('┃'), Cell('┃')],
    [Cell('┗'), Cell('┛')],
  ]);

  // Copy it into a writable buffer.
  final buffer = WritableBuffer(5, 5)..fillFrom(other, x: 2, y: 1);
  ```

[`bitblt`]: https://en.wikipedia.org/wiki/Bit_blit

## 0.3.1

Applied basic optimizations to ANSI escape code screens.

```dart
// Before
screen.print('GREEN', 0, 2, foreground: green, background: blue);
screen.update();
```

... produced:

```txt
\x1B[2J\n
\x1B[0m \x1B[0m \x1B[0m \x1B[0m \x1B[0m\n
\x1B[0mH\x1B[0mE\x1B[0mL\x1B[0mL\x1B[0mO\n
      \x1B[38;2;0;255;0m\x1B[48;2;0;0;255mG\x1B[38;2;0;255;0m\x1B[48;2;0;0;255mR\x1B[38;2;0;255;0m\x1B[48;2;0;0;255mE\x1B[38;2;0;255;0m\x1B[48;2;0;0;255mE\x1B[38;2;0;255;0m\x1B[48;2;0;0;255mN\n
```

... and after the optimizations:

```txt
\x1B[2J\n
      \n
\x1B[0m\x1B[38;2;0;255;0m\x1B[48;2;0;0;255mGREEN\n
```

## 0.3.0

In accordance with our [design](DESIGN.md), any features that are not
appropriate for a high-level _canvas-like_ API were removed from this package:

- `Screen(framesPerSecond: ...)` and `Screen.terminal(framesPerSecond: ...)`.
- `<Screen>.onFrame`:

  ```dart
  // Before
  final screen = Screen(framesPerSecond: 30);
  screen.onFrame.listen((_) { /* ... */ })

  // After
  final screen = Screen();
  final frames = Stopwatch()..start();
  Stream.periodic(Duration(milliseconds: 1000 ~/ 30)).listen((_) {
    final time = frames.elapsed;
    frames.reset();
    /* ... */
  });
  ```

- `Terminal.usingAnsiStdio({stdin: ...})`:

  ```dart
  // Before
  import 'package:griddle/griddle.dart';

  void main() {
    Screen.terminal(Terminal.usingAnsiStdio());
  }
  ```

  ```dart
  // After
  import 'dart:io';

  import 'package:griddle/griddle.dart';

  void main() {
    Screen.display(
      Display.fromAnsiTerminal(
        stdout,
        width: () => stdout.width,
        height: () => stdout.height,
      ),
    );
  }
  ```

These changes allow us to focus on just pushing pixels and output, versus
worrying about other elements of UI, such as the update loop or user input,
which are better suited to other packages, as well as keeping this package
completely platform agnostic.

**If this API remains relatively stable, it will (eventually) become `1.0.0`.**

## 0.2.0

New release with many bug fixes, changes, and [new examples](example/README.md)!

- Added `Buffer.fromCells` and `Buffer.fromMatrix` as helpful factory methods.
- Added `<Buffer>.length` as an alias for `<Buffer>.width * <Buffer>.height`.
- Added `<Buffer>[]` and `<Buffer>[]=` for indexed reads/writes into a buffer.
- Added `<Buffer>.toList()` which returns a copy of the underlying cells.
- Added `<Buffer>.toMatrix()` which returns a copy of cells as nested lists.
- Changed `Buffer(width, height)`, added an optional `initialCell` parameter:

  ```dart
  // Before.
  Buffer(3, 3);

  // Semantically identical.
  Buffer(3, 3, initialCell: Cell.blank);

  // Newly possible.
  Buffer(3, 3, initialCell: Cell('X'));
  ```

- Terminal output always skips line `0` and starts on line `1` for readability.
- Fixed a bug where cells were stored as `y * height + x`, not `y * width + x`.
- Fixed a bug where `<Buffer>.clear()` just didn't work, period.
- Fixed a bug where `<Buffer>.resize()` created an invalid state.
- Changed `<Buffer>.resize()` to have named parameters, and added `expand`:

  ```dart
  // Before
  buffer.resize(3, 4);

  // Semantically identical.
  buffer.resize(width: 3, height: 4);

  // Newly possible: just change width OR heght.
  buffer.resize(width: 3);
  buffer.resize(height: 4);

  // Newly possible: specify what cell to fill when expanding the buffer.
  // (Defaults to Cell.blank)
  buffer.resize(width: 3, height: 4, expand: Cell('X'));
  ```

- Changed `<Buffer>.fill()` to have only named parameters:

  ```dart
  // Before
  buffer.fill(1, 1, 2, 2);

  // Semantically identical.
  buffer.fill(x: 1, y: 1, width: 2, height: 2);
  ```

## 0.1.0

- Initial (real) commit, with a simple API and animated text example.

## 0.0.0

- Initial release as a placeholder only.


example/fun_text.dart
import 'dart:io' show stdout;
import 'dart:math' as math;

import 'package:griddle/griddle.dart';

/// A sample application, similar to the one provided in [termpixels][].
///
/// [termpixels]: https://github.com/loganzartman/termpixels
void main() {
  run(
    Screen.display(
      Display.fromAnsiTerminal(
        stdout,
        width: () => stdout.terminalColumns,
        height: () => stdout.terminalLines,
      ),
    ),
  );
}

void run(Screen screen) {
  const string = 'Hello World, from Griddle for Dart!';
  final frames = Stopwatch()..start();
  Stream<void>.periodic(const Duration(milliseconds: 1000 ~/ 30)).listen((_) {
    screen.clear();

    for (var i = 0; i < string.length; i++) {
      final t = DateTime.now().millisecondsSinceEpoch / 1000;
      final f = i / string.length;
      final c = Color.fromHSL(f * 300 + t, 1, 0.5);
      final x = screen.width ~/ 2 - string.length ~/ 2;
      final o = math.sin(t * 3 + f * 5) * 2;
      final y = (screen.height / 2 + o).round();

      screen.print(string[i], x + i, y, foreground: c);
    }

    screen
      ..print('${frames.elapsedMilliseconds}ms', 0, screen.height ~/ 2 - 2)
      ..update();

    frames.reset();
  });
}
